File: python/agentseal/notify.py
# agentseal/notify.py
"""
Desktop notifications for AgentSeal Shield.

Uses OS built-in notification mechanisms - no additional dependencies.
macOS: osascript, Linux: notify-send, Fallback: terminal bell + stderr.
"""

import platform
import subprocess
import sys
import time
from typing import Optional


_SEVERITY_ICONS = {
    "critical": "CRITICAL",
    "high": "HIGH",
    "medium": "MEDIUM",
    "low": "LOW",
}


class Notifier:
    """Cross-platform desktop notification dispatcher with throttling."""

    def __init__(self, enabled: bool = True, min_interval: float = 30.0):
        self._enabled = enabled
        self._min_interval = min_interval
        self._last_notify_time: float = 0.0
        self._platform = platform.system()

    def notify(
        self,
        title: str,
        message: str,
        *,
        urgent: bool = False,
    ) -> bool:
        """Send a desktop notification. Returns True if notification was sent.

        Respects throttle interval - returns False if called too soon after
        the previous notification.
        """
        if not self._enabled:
            return False

        now = time.monotonic()
        if now - self._last_notify_time < self._min_interval:
            return False

        sent = self._dispatch(title, message, urgent=urgent)
        if sent:
            self._last_notify_time = now
        return sent

    def notify_threat(
        self,
        item_name: str,
        item_type: str,
        severity: str,
        detail: str,
    ) -> bool:
        """Send a threat notification with standard formatting."""
        level = _SEVERITY_ICONS.get(severity, severity.upper())
        title = f"AgentSeal Shield - {level}"
        message = f"{item_type}: {item_name}\n{detail}"
        return self.notify(title, message, urgent=severity in ("critical", "high"))

    def _dispatch(self, title: str, message: str, *, urgent: bool = False) -> bool:
        """Send notification via platform-specific mechanism."""
        if self._platform == "Darwin":
            return self._notify_macos(title, message, urgent=urgent)
        elif self._platform == "Linux":
            return self._notify_linux(title, message, urgent=urgent)
        return self._notify_fallback(title, message)

    def _notify_macos(self, title: str, message: str, *, urgent: bool = False) -> bool:
        """macOS notification via osascript."""
        # Escape double quotes for AppleScript
        safe_title = title.replace('"', '\\"')
        safe_message = message.replace('"', '\\"').replace("\n", " - ")
        sound = ' sound name "Basso"' if urgent else ""
        script = (
            f'display notification "{safe_message}" '
            f'with title "{safe_title}"{sound}'
        )
        try:
            subprocess.run(
                ["osascript", "-e", script],
                capture_output=True,
                timeout=5,
            )
            return True
        except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
            return self._notify_fallback(title, message)

    def _notify_linux(self, title: str, message: str, *, urgent: bool = False) -> bool:
        """Linux notification via notify-send."""
        urgency = "critical" if urgent else "normal"
        try:
            subprocess.run(
                [
                    "notify-send",
                    title,
                    message,
                    f"--urgency={urgency}",
                    "--icon=dialog-warning",
                ],
                capture_output=True,
                timeout=5,
            )
            return True
        except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
            return self._notify_fallback(title, message)

    @staticmethod
    def _notify_fallback(title: str, message: str) -> bool:
        """Fallback: terminal bell + colored stderr output."""
        sys.stderr.write(f"\a\033[93m[{title}]\033[0m {message}\n")
        sys.stderr.flush()
        return True

File: js/src/notify.ts
/**
 * Desktop notifications for AgentSeal Shield.
 *
 * Uses OS built-in notification mechanisms — no additional dependencies.
 * macOS: osascript, Linux: notify-send, Fallback: terminal bell + stderr.
 *
 * Port of Python agentseal/notify.py.
 */

import { execFileSync } from "node:child_process";
import { platform } from "node:os";

const SEVERITY_ICONS: Record<string, string> = {
  critical: "CRITICAL",
  high: "HIGH",
  medium: "MEDIUM",
  low: "LOW",
};

export class Notifier {
  private _enabled: boolean;
  private _minInterval: number;
  private _lastNotifyTime: number = -Infinity;
  private _platform: string;

  constructor(enabled = true, minInterval = 30.0) {
    this._enabled = enabled;
    this._minInterval = minInterval;
    this._platform = platform();
  }

  get enabled(): boolean {
    return this._enabled;
  }

  /** Send a desktop notification. Returns true if sent. Respects throttle interval. */
  notify(title: string, message: string, urgent = false): boolean {
    if (!this._enabled) return false;

    const now = performance.now() / 1000;
    if (now - this._lastNotifyTime < this._minInterval) return false;

    const sent = this._dispatch(title, message, urgent);
    if (sent) this._lastNotifyTime = now;
    return sent;
  }

  /** Send a threat notification with standard formatting. */
  notifyThreat(
    itemName: string,
    itemType: string,
    severity: string,
    detail: string,
  ): boolean {
    const level = SEVERITY_ICONS[severity] ?? severity.toUpperCase();
    const title = `AgentSeal Shield - ${level}`;
    const message = `${itemType}: ${itemName}\n${detail}`;
    return this.notify(title, message, severity === "critical" || severity === "high");
  }

  private _dispatch(title: string, message: string, urgent: boolean): boolean {
    if (this._platform === "darwin") return this._notifyMacOS(title, message, urgent);
    if (this._platform === "linux") return this._notifyLinux(title, message, urgent);
    return this._notifyFallback(title, message);
  }

  private _notifyMacOS(title: string, message: string, urgent: boolean): boolean {
    const safeTitle = title.replace(/"/g, '\\"');
    const safeMessage = message.replace(/"/g, '\\"').replace(/\n/g, " - ");
    const sound = urgent ? ' sound name "Basso"' : "";
    const script = `display notification "${safeMessage}" with title "${safeTitle}"${sound}`;
    try {
      execFileSync("osascript", ["-e", script], { timeout: 5000, stdio: "pipe" });
      return true;
    } catch {
      return this._notifyFallback(title, message);
    }
  }

  private _notifyLinux(title: string, message: string, urgent: boolean): boolean {
    const urgency = urgent ? "critical" : "normal";
    try {
      execFileSync(
        "notify-send",
        [title, message, `--urgency=${urgency}`, "--icon=dialog-warning"],
        { timeout: 5000, stdio: "pipe" },
      );
      return true;
    } catch {
      return this._notifyFallback(title, message);
    }
  }

  private _notifyFallback(title: string, message: string): boolean {
    process.stderr.write(`\x07\x1b[93m[${title}]\x1b[0m ${message}\n`);
    return true;
  }
}
